src/MongoDB.Driver.Core/Core/Authentication/SaslAuthenticator.cs

```diff
@@ -65,12 +65,16 @@ public void Authenticate(IConnection connection, ConnectionDescription descripti
             Ensure.IsNotNull(connection, nameof(connection));
             Ensure.IsNotNull(description, nameof(description));
 
+            var firstStepResult = description.IsMasterResult.SpeculativeAuthenticate;
             using (var conversation = new SaslConversation(description.ConnectionId))
             {
-                var currentStep = _mechanism.Initialize( connection, conversation, description);
+                 var firstStepAndCommand = CreateFirstStepAndCommand(_mechanism, connection, conversation, description);
+                /* We cannot skip the cast to StepAndCommand below without adding a reference to ValueTuple */
+                var (currentStep, command) = (StepAndCommand)(firstStepResult  == null
+                    ? firstStepAndCommand
+                    : CreateNextStepAndCommand(conversation, firstStepAndCommand.Step, firstStepResult));
 
-                var command = CreateStartCommand(currentStep);
-                while (true)
+                while (currentStep != null)
```

Comment:
Since the previous thread (https://github.com/vincentkam/mongo-csharp-driver/pull/95/#discussion_r401004782) was "oudated" by my changes, I'm starting a new thread here, with a copy of my last comment:


My initial design allowed Authentication to be continued at ANY step, which is probably overkill for this PR.
So I agree with @rstam and @DmitryLukyanov that resulted in an unneeded overload that was actually probably should have been a separate method named "ResumeAuthenticate".

@rstam : I do believe your refactor is correct, however, I find the need to check if command is null every single iteration of the loop to be somewhat confusing (especially since such a condition should really only happen once.

@rstam @DmitryLukyanov : I've refactored my PR and pushed the changes: What do you think of this new refactored version? I think the key feature of my refactor is that the loop feels simpler to me: all the special case logic happens prior to the loop starting: the loop simply executes the current command and creates the next command.

Reply:
The current version of the refactored loop also makes it so that we don't need to store an extra bool flag or store a null in `command` to act as a flag. We loop until `currentStep` is null and that's it, and to me, that's a much easier loop to understand.

Reply:
I've taken Robert's suggested offline refactor that's combined elements of my refactor with his and @DmitryLukyanov 's :)